lib/common/xml.c

```diff
@@ -900,6 +901,54 @@ pcmk__xml_is_name_char(const char *utf8, int *len)
            || ((c >= 0x10000) && (c <= 0xEFFFF));
 }
 
+/*!
+ * \internal
+ * \brief Sanitize a string so it is usable as an XML ID
+ *
+ * An ID must match the Name production as defined here:
+ * https://www.w3.org/TR/xml/#NT-Name.
+ *
+ * Convert an invalid start character to \c '_'. Convert an invalid character
+ * after the start character to \c '.'.
+ *
+ * \param[in,out] id  String to sanitize
+ */
+void
+pcmk__xml_sanitize_id(char *id)
+{
+    bool valid = true;
+    int len = 0;
+
+    // If id is empty or NULL, there's no way to make it a valid XML ID
+    CRM_ASSERT(!pcmk__str_empty(id));
+
+    /* @TODO Suppose there are two strings and each has an invalid ID character
+     * in the same position. The strings are otherwise identical. Both strings
+     * will be sanitized to the same valid ID, which is incorrect.
+     *
+     * To avoid this, one option might be to append a UUID string to sanitized
+     * IDs.
+     */
```

Comment:
Just note that the caller is responsible for ensuring the sanitized ID does not exist in the XML document before using it if uniqueness is desired. (Actually doing that at the callers can wait ...)